hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/router/RouterAsyncSnapshot.java

```diff
@@ -0,0 +1,204 @@
+import org.apache.hadoop.hdfs.protocol.SnapshottableDirectoryStatus;
+import org.apache.hadoop.hdfs.server.federation.resolver.ActiveNamenodeResolver;
+import org.apache.hadoop.hdfs.server.federation.resolver.FederationNamespaceInfo;
+import org.apache.hadoop.hdfs.server.federation.resolver.RemoteLocation;
+import org.apache.hadoop.hdfs.server.federation.router.async.ApplyFunction;
+import org.apache.hadoop.hdfs.server.federation.router.async.AsyncUtil;
+import org.apache.hadoop.hdfs.server.namenode.NameNode;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static org.apache.hadoop.hdfs.server.federation.router.async.AsyncUtil.asyncApply;
+import static org.apache.hadoop.hdfs.server.federation.router.async.AsyncUtil.asyncReturn;
+
+/**
+ * Module that implements all the asynchronous RPC calls related to snapshots in
+ * {@link ClientProtocol} in the {@link RouterRpcServer}.
+ */
+public class RouterAsyncSnapshot extends RouterSnapshot {
+  /** RPC server to receive client calls. */
+  private final RouterRpcServer rpcServer;
+  /** RPC clients to connect to the Namenodes. */
+  private final RouterRpcClient rpcClient;
+  /** Find generic locations. */
+  private final ActiveNamenodeResolver namenodeResolver;
+
+  public RouterAsyncSnapshot(RouterRpcServer server) {
+    super(server);
+    this.rpcServer = server;
+    this.rpcClient = this.rpcServer.getRPCClient();
+    this.namenodeResolver = rpcServer.getNamenodeResolver();
+  }
+
+  @Override
+  public String createSnapshot(String snapshotRoot, String snapshotName) throws IOException {
+    rpcServer.checkOperation(NameNode.OperationCategory.WRITE);
+
+    final List<RemoteLocation> locations =
+        rpcServer.getLocationsForPath(snapshotRoot, true, false);
+    RemoteMethod method = new RemoteMethod("createSnapshot",
+        new Class<?>[] {String.class, String.class}, new RemoteParam(),
+        snapshotName);
+
+    if (rpcServer.isInvokeConcurrent(snapshotRoot)) {
+      rpcClient.invokeConcurrent(locations, method, String.class);
+      asyncApply((ApplyFunction<Map<RemoteLocation, String>, String>)
+          results -> {
+          Map.Entry<RemoteLocation, String> firstelement =
+              results.entrySet().iterator().next();
+          RemoteLocation loc = firstelement.getKey();
+          String result = firstelement.getValue();
+          return result.replaceFirst(loc.getDest(), loc.getSrc());
+        });
+    } else {
+      rpcClient.invokeSequential(method, locations, String.class, null);
+      asyncApply((ApplyFunction<RemoteResult<RemoteLocation, String>, String>)
+          response -> {
+          RemoteLocation loc = response.getLocation();
+          String invokedResult = response.getResult();
+          return invokedResult.replaceFirst(loc.getDest(), loc.getSrc());
+        });
+    }
+    return AsyncUtil.asyncReturn(String.class);
```

Comment:
@LeoLeeeeee Hi, thank you for your contribution. I think AsyncUtil can be deleted here. You can use asyncReturn(String.class) directly.